### PR TITLE
Allow mesh generators during recover

### PR DESCRIPTION
**Pull request by hugary1995**
I am adding a parameter called `use_checkpoint_mesh` which defaults to true in the Mesh block. The parameter should be self-explanatory.

This is a hacky fix for #21161. This is obviously not the long-term solution, but there is also nothing wrong with this new parameter itself.

### CODE REVIEW

**Comment by hugary1995** [comment]
@jiangwen84

**Comment by hugary1995** [comment]
hmm I don't have the privilege to add the recover sweeps. Can someone add those for me please? :)

File: framework/src/actions/ExecuteMeshGenerators.C [hunk 1]
```diff
@@ -27,7 +27,8 @@ ExecuteMeshGenerators::act()
   // Don't do mesh generators when recovering as the master app or using master mesh! We do need
```
Comment (GiudGiud) [inline framework/src/actions/ExecuteMeshGenerators.C:27]: needs to be updated

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -156,9 +161,9 @@ MooseMesh::validParams()
   params.registerBase("MooseMesh");
 
   // groups
-  params.addParamNamesToGroup(
-      "dim nemesis patch_update_strategy construct_node_list_from_side_list patch_size",
-      "Advanced");
+  params.addParamNamesToGroup("dim nemesis patch_update_strategy "
+                              "construct_node_list_from_side_list patch_size use_checkpoint_mesh",
+                              "Advanced");
```
Comment (GiudGiud) [inline framework/src/mesh/MooseMesh.C:166]: it s time someone made real parameter groups here. I ll do it if you dont

**Review by GiudGiud — CHANGES_REQUESTED** [review]
What if the number of elements change during the mesh generation step?

File: framework/src/actions/ExecuteMeshGenerators.C (hunk 1 above)
Comment (hugary1995) [reply framework/src/actions/ExecuteMeshGenerators.C:27]: :+1:

File: framework/src/mesh/MooseMesh.C (hunk 2 above)
Comment (hugary1995) [reply framework/src/mesh/MooseMesh.C:166]: I'll leave it to you as I don't fully understand some of those advanced parameters.

**Comment by hugary1995** [comment]
> What if the number of elements change during the mesh generation step?

So `use_checkpoint_mesh = false` should only be set when the user knows that the mesh didn't change in the prior simulation. I'll add this to both the parameter description and the documentation.

**Comment by GiudGiud** [comment]
what kind of error do you get if the user is being reckless here? Just want to be sure it s not a seg fault.

**Comment by hugary1995** [comment]
Hmm, good question. I think very likely what will happen is that during dataLoad of some material properties, some locations will be nullptr if the mesh is different from the checkpoint mesh, hence, unfortunately, a segfault.

But in that case, the right fix is not to check if the new mesh is the same as the checkpoint mesh. Instead, we should enable better error handling in DataIO. That's probably out of the scope of this PR.

**Comment by GiudGiud** [comment]
New seg faults are automatically in-scope in my opinion. Let's try though, maybe the situation is not that gloomy

**Comment by hugary1995** [comment]
lol okay, I'll take a look.

**Comment by hugary1995** [comment]
Indeed, this is way easier than I expected. When flagging elements for adaptivity during recover, the marker values will be invalid if the meshes differ, and such case has already been handled with a proper mooseError. I also added two regression tests to test that.

**Comment by hugary1995** [comment]
hmmm, I guess I didn't write the `restart/regenerate_mesh.adaptivity_error` regression test properly...

**Comment by GiudGiud** [comment]
Hey @hugary1995 could you please document why you are closing this? Just so we know if this gets revived by someone else needing it

**Comment by hugary1995** [comment]
Sorry I was going to leave a message here but got distracted.

In short, I found an existing capability that is equivalent to the workaround proposed in this PR. So I believe this PR is no longer relevant. Although the reference issue #21161 remains.

The existing capability I found is the so-called "advanced restart" option which can be used as
```python
[Problem]
  restart_file_base = 'cp/LATEST'
[]
```
This option loads all variables, stateful properties, etc. from the checkpoint, while still provides the flexibility to re-generate the mesh.

**Pull request closed without merging**